File: src/org/cerion/symcalc/expression/function/Divide.java
package org.cerion.symcalc.expression.function;

import org.cerion.symcalc.expression.ErrorExpr;
import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.NumberExpr;

public class Divide extends FunctionExpr {

	public Divide(Expr ...e) {
		super(FunctionType.DIVIDE, e);
	}

	@Override
	protected Expr evaluate() {
		
		if(size() != 2)
			return new ErrorExpr("invalid parameters");
		
		Expr a = get(0);
		Expr b = get(1);
		
		//Identity
		if(b.isNumber() && ((NumberExpr)b).isOne())
			return a;
				
		if(a.isNumber() && b.isNumber())
		{
			NumberExpr n1 = (NumberExpr)a;
			NumberExpr n2 = (NumberExpr)b;
			
			return n1.divide(n2);
		}

		return this;
	}
	
	@Override
	public String toString()
	{
		if(size() == 2)
			return get(0) + " / " + get(1);
		
		return super.toString();
	}
}

File: src/org/cerion/symcalc/expression/function/arithmetic/Divide.kt
package org.cerion.symcalc.expression.function.arithmetic

import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.FunctionExpr

class Divide(vararg e: Expr) : FunctionExpr(FunctionExpr.FunctionType.DIVIDE, *e) {

    override fun evaluate(): Expr {
        val a = get(0)
        val b = get(1)

        //Identity
        if (b.isNumber && b.asNumber().isOne)
            return a

        if (a.isNumber && b.isNumber)
            return a.asNumber() / b.asNumber()

        return this
    }

    override fun toString(): String = if (size() == 2) get(0).toString() + " / " + get(1) else super.toString()

    override fun validate() {
        validateParameterCount(2)
    }
}
